File: .skills/openclaw-skills/skills/shifenghu/aicade-galaxy-skills/scripts/export_artifact.py
#!/usr/bin/env python3

import json
import sys
import urllib.error
import urllib.request
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List, Optional


DEFAULT_OUTPUT_DIR = "output"
OUTPUT_FILE_NAME = "aicade-galaxy-skill.json"
DEFAULT_SERVICES_PATH = "/admin/gateway/services"


def main() -> None:
    try:
        config = load_config()
        services = discover_services(config)
        artifact = build_artifact(config, services)
        output_path = Path(config["output_path"])
        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_text(
            json.dumps(artifact, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
        )
        print(f"artifact written: {output_path}")
        print(
            json.dumps(
                {
                    "baseUrl": config["base_url"],
                    "toolCount": artifact["toolCount"],
                    "toolNames": [tool["name"] for tool in artifact["tools"]],
                },
                ensure_ascii=False,
                indent=2,
            )
        )
    except Exception as exc:  # noqa: BLE001
        print("Failed to export AICADE Galaxy artifact.", file=sys.stderr)
        print(str(exc), file=sys.stderr)
        if is_auth_like_error(str(exc)):
            print(
                "Check whether AICADE_GALAXY_API_KEY is configured correctly in .env.",
                file=sys.stderr,
            )
        sys.exit(1)


def load_config() -> Dict[str, str]:
    env = load_env(Path(".env"))

    base_url = env.get("AICADE_GALAXY_BASE_URL") or env.get("CLAWHUB_BASE_URL")
    api_key = env.get("AICADE_GALAXY_API_KEY")
    output_dir = env.get("AICADE_GALAXY_OUTPUT_PATH") or DEFAULT_OUTPUT_DIR
    debug = is_debug_enabled(env.get("AICADE_GALAXY_DEBUG"))

    if not base_url:
        raise RuntimeError(
            'Missing required environment variable: AICADE_GALAXY_BASE_URL. Run "node scripts/setup_env.mjs" first, or fall back to "python3 scripts/setup_env.py", or update .env.'
        )
    if not api_key:
        raise RuntimeError(
            'Missing required environment variable: AICADE_GALAXY_API_KEY. Run "node scripts/setup_env.mjs" first, or fall back to "python3 scripts/setup_env.py", or update .env.'
        )

    return {
        "base_url": base_url.rstrip("/"),
        "services_path": DEFAULT_SERVICES_PATH,
        "output_path": str(Path(output_dir) / OUTPUT_FILE_NAME),
        "api_key": api_key,
        "debug": debug,
    }


def load_env(path: Path) -> Dict[str, str]:
    if not path.exists():
        return {}

    result: Dict[str, str] = {}
    for raw_line in path.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        result[key.strip()] = strip_quotes(value.strip())
    return result


def strip_quotes(value: str) -> str:
    if len(value) >= 2 and value[0] == value[-1] and value[0] in {"'", '"'}:
        return value[1:-1]
    return value


def discover_services(config: Dict[str, str]) -> List[dict]:
    url = join_url(config["base_url"], config["services_path"])
    headers = {
        "Accept": "application/json",
        "X-API-Key": config["api_key"],
    }
    log_debug(
        config,
        {
            "event": "request",
            "method": "GET",
            "url": url,
            "headers": headers,
        },
    )
    request = urllib.request.Request(url, headers=headers, method="GET")

    try:
        with urllib.request.urlopen(request) as response:
            body = response.read().decode("utf-8", errors="replace")
            content_type = response.headers.get("Content-Type", "")
            status_code = response.status
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(
            f"Failed to discover services: HTTP {exc.code}. "
            f"Content-Type={exc.headers.get('Content-Type', '')}. "
            f"Body preview={preview_text(body)}"
        ) from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"Failed to discover services: {exc.reason}") from exc

    log_debug(
        config,
        {
            "event": "response",
            "method": "GET",
            "url": url,
            "statusCode": status_code,
            "contentType": content_type,
            "bodyPreview": preview_text(body),
        },
    )

    try:
        payload = json.loads(body)
    except json.JSONDecodeError as exc:
        raise RuntimeError(
            f"Failed to parse services response as JSON. "
            f"Content-Type={content_type}. "
            f"Body preview={preview_text(body)}"
        ) from exc

    if not isinstance(payload, dict) or not isinstance(payload.get("data"), list):
        raise RuntimeError("Invalid services response shape")

    return payload["data"]


def build_artifact(config: Dict[str, str], services: List[dict]) -> dict:
    tools = []
    for service in services:
        tool = build_tool(service)
        if tool:
            tools.append(tool)

    return {
        "version": "1.0",
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "baseUrl": config["base_url"],
        "toolCount": len(tools),
        "tools": tools,
    }


def build_tool(service: dict) -> Optional[dict]:
    if not service.get("enabled", False):
        return None

    input_schema = service.get("inputSchema") or {}
    method = str(input_schema.get("method", "")).upper()
    if method not in {"GET", "POST"}:
        return None

    properties = input_schema.get("properties")
    if not isinstance(properties, dict):
        properties = {}

    required = input_schema.get("required")
    if not isinstance(required, list):
        required = []

    output_schema = service.get("outputSchema") or {}
    response_properties = output_schema.get("properties")
    if not isinstance(response_properties, dict):
        response_properties = {}

    response_fields = list_response_paths(response_properties)

    return {
        "name": create_tool_name(
            str(service.get("serviceId", service.get("name", "service")))
        ),
        "description": build_tool_description(
            service, method, properties, required, response_fields
        ),
        "inputSchema": {
            "type": "object",
            "properties": {
                **properties,
                "responsePaths": build_response_paths_property(response_fields),
            },
            "required": [key for key in required if key in properties],
            "additionalProperties": False,
        },
        "metadata": {
            "serviceId": service.get("serviceId"),
            "title": service.get("title")
            or service.get("serviceName")
            or service.get("name"),
            "method": method,
            "path": service.get("routePath"),
            "authentication": {
                "type": "apiKeyHeader",
                "required": True,
                "headerName": "X-API-Key",
                "envName": "AICADE_GALAXY_API_KEY",
            },
            "responseFields": response_fields,
            "invocationProtocol": {
                "reservedFields": ["responsePaths"],
                "responsePathsField": "responsePaths",
            },
        },
    }


def create_tool_name(service_id: str) -> str:
    return "".join(ch if ch.isalnum() or ch == "_" else "_" for ch in service_id)


def build_tool_description(
    service: dict,
    method: str,
    properties: dict,
    required: list,
    response_fields: List[str],
) -> str:
    parts = [
        service.get("title") or service.get("serviceName") or service.get("name") or "AICADE service",
        service.get("description") or "",
        "Use query parameters and expect a JSON response." if method == "GET" else "Use a JSON request body and expect a JSON response.",
    ]

    if properties:
        names = list(properties.keys())
        parts.append(f"Parameters: {', '.join(names)}.")

        required_names = [name for name in required if name in properties]
        optional_names = [name for name in names if name not in required_names]
        if required_names:
            parts.append(f"Required parameters: {', '.join(required_names)}.")
        if optional_names:
            parts.append(f"Optional parameters: {', '.join(optional_names)}.")

    if response_fields:
        parts.append(
            f"Optional reserved field: responsePaths. Selectable response fields: {', '.join(response_fields)}. Use responsePaths to request only the fields you need."
        )
    else:
        parts.append(
            "Optional reserved field: responsePaths. Response selection is supported, but this service does not declare output fields."
        )

    return " ".join(part for part in parts if part)


def build_response_paths_property(response_fields: List[str]) -> dict:
    description = "Optional. Select one or more response field paths to return."
    if response_fields:
        description = (
            f"{description} Available values: {', '.join(response_fields)}."
        )

    return {
        "type": "array",
        "description": description,
        "items": {
            "type": "string",
        },
    }


def list_response_paths(properties: dict) -> List[str]:
    paths: List[str] = []
    for key, prop in properties.items():
        walk_schema_property(key, prop, paths)
    return paths


def walk_schema_property(current_path: str, prop: dict, paths: List[str]) -> None:
    paths.append(current_path)

    if prop.get("type") == "object" and isinstance(prop.get("properties"), dict):
        for key, child in prop["properties"].items():
            walk_schema_property(f"{current_path}.{key}", child, paths)

    if prop.get("type") == "array" and isinstance(prop.get("items"), dict):
        paths.append(f"{current_path}[]")
        items = prop["items"]
        if items.get("type") == "object" and isinstance(
            items.get("properties"), dict
        ):
            for key, child in items["properties"].items():
                walk_schema_property(f"{current_path}[].{key}", child, paths)


def join_url(base_url: str, path: str) -> str:
    normalized_path = path[1:] if path.startswith("/") else path
    return f"{base_url}/{normalized_path}"


def is_auth_like_error(message: str) -> bool:
    return (
        "401" in message
        or "403" in message
        or "X-API-Key" in message
        or "AICADE_GALAXY_API_KEY" in message
    )


def is_debug_enabled(value: Optional[str]) -> bool:
    return bool(value) and value.lower() == "true"


def log_debug(config: Dict[str, str], payload: dict) -> None:
    if not config.get("debug"):
        return
    print(
        f"[AICADE_GALAXY_DEBUG] {json.dumps(payload, ensure_ascii=False, indent=2)}"
    )


def preview_text(text: str, limit: int = 240) -> str:
    compact = " ".join(text.split())
    if len(compact) <= limit:
        return compact
    return compact[:limit] + "..."


if __name__ == "__main__":
    main()

File: .skills/openclaw-skills/skills/shifenghu/aicade-galaxy-skill/scripts/export_artifact.mjs
#!/usr/bin/env node

import { mkdir, readFile, writeFile } from "node:fs/promises";
import http from "node:http";
import https from "node:https";

const DEFAULT_OUTPUT_DIR = "output";
const OUTPUT_FILE_NAME = "aicade-galaxy-skill.json";
const DEFAULT_SERVICES_PATH = "/admin/gateway/services";

async function main() {
  try {
    const config = await loadConfig();
    const services = await discoverServices(config);
    const artifact = buildArtifact(config, services);

    const outputPath = config.outputPath;
    await mkdir(dirname(outputPath), { recursive: true });
    await writeFile(
      outputPath,
      `${JSON.stringify(artifact, null, 2)}\n`,
      "utf8",
    );

    console.log(`artifact written: ${outputPath}`);
    console.log(
      JSON.stringify(
        {
          baseUrl: config.baseUrl,
          toolCount: artifact.toolCount,
          toolNames: artifact.tools.map((tool) => tool.name),
        },
        null,
        2,
      ),
    );
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error("Failed to export AICADE Galaxy artifact.");
    console.error(message);
    if (isAuthLikeError(message)) {
      console.error(
        "Check whether AICADE_GALAXY_API_KEY is configured correctly in .env.",
      );
    }
    process.exitCode = 1;
  }
}

async function loadConfig() {
  const env = await loadEnv(".env");

  const baseUrl = env.AICADE_GALAXY_BASE_URL || env.CLAWHUB_BASE_URL;
  const apiKey = env.AICADE_GALAXY_API_KEY;
  const outputDir = env.AICADE_GALAXY_OUTPUT_PATH || DEFAULT_OUTPUT_DIR;
  const debug = isDebugEnabled(env.AICADE_GALAXY_DEBUG);

  if (!baseUrl) {
    throw new Error(
      'Missing required environment variable: AICADE_GALAXY_BASE_URL. Run "node scripts/setup_env.mjs" or "python3 scripts/setup_env.py", or update .env.',
    );
  }

  if (!apiKey) {
    throw new Error(
      'Missing required environment variable: AICADE_GALAXY_API_KEY. Run "node scripts/setup_env.mjs" or "python3 scripts/setup_env.py", or update .env.',
    );
  }

  return {
    baseUrl: baseUrl.replace(/\/+$/, ""),
    servicesPath: DEFAULT_SERVICES_PATH,
    outputPath: joinPath(outputDir, OUTPUT_FILE_NAME),
    apiKey,
    debug,
  };
}

async function loadEnv(path) {
  try {
    const content = await readFile(path, "utf8");
    return parseEnv(content);
  } catch {
    return {};
  }
}

function parseEnv(content) {
  const result = {};

  for (const rawLine of content.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) {
      continue;
    }

    const equalsIndex = line.indexOf("=");
    if (equalsIndex <= 0) {
      continue;
    }

    const key = line.slice(0, equalsIndex).trim();
    const value = line.slice(equalsIndex + 1).trim();
    result[key] = stripQuotes(value);
  }

  return result;
}

function stripQuotes(value) {
  if (
    (value.startsWith('"') && value.endsWith('"')) ||
    (value.startsWith("'") && value.endsWith("'"))
  ) {
    return value.slice(1, -1);
  }

  return value;
}

async function discoverServices(config) {
  const url = joinUrl(config.baseUrl, config.servicesPath);
  const headers = {
    Accept: "application/json",
    "X-API-Key": config.apiKey,
  };
  logDebug(config, {
    event: "request",
    method: "GET",
    url,
    headers,
  });
  const response = await requestJson(url, headers);
  logDebug(config, {
    event: "response",
    method: "GET",
    url,
    statusCode: response.statusCode,
    contentType: response.contentType,
    bodyPreview: previewText(response.body),
  });

  let payload;
  try {
    payload = JSON.parse(response.body);
  } catch (error) {
    throw new Error(
      `Failed to parse services response as JSON. Content-Type=${response.contentType}. Body preview=${previewText(response.body)}`,
    );
  }

  if (!payload || typeof payload !== "object" || !Array.isArray(payload.data)) {
    throw new Error("Invalid services response shape");
  }

  return payload.data;
}

function requestJson(urlString, headers) {
  return new Promise((resolve, reject) => {
    const url = new URL(urlString);
    const transport = url.protocol === "https:" ? https : http;
    const request = transport.request(
      url,
      {
        method: "GET",
        headers,
      },
      (response) => {
        const chunks = [];

        response.on("data", (chunk) => {
          chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
        });

        response.on("end", () => {
          const body = Buffer.concat(chunks).toString("utf8");
          const contentType = String(response.headers["content-type"] || "");
          const statusCode = response.statusCode || 0;

          if (statusCode >= 400) {
            reject(
              new Error(
                `Failed to discover services: HTTP ${statusCode}. Content-Type=${contentType}. Body preview=${previewText(body)}`,
              ),
            );
            return;
          }

          resolve({ body, contentType, statusCode });
        });
      },
    );

    request.on("error", (error) => {
      reject(new Error(`Failed to discover services: ${error.message}`));
    });

    request.end();
  });
}

function buildArtifact(config, services) {
  const tools = [];
  for (const service of services) {
    const tool = buildTool(service);
    if (tool) {
      tools.push(tool);
    }
  }

  return {
    version: "1.0",
    generatedAt: new Date().toISOString(),
    baseUrl: config.baseUrl,
    toolCount: tools.length,
    tools,
  };
}

function buildTool(service) {
  if (!service?.enabled) {
    return null;
  }

  const inputSchema = service.inputSchema || {};
  const method = String(inputSchema.method || "").toUpperCase();
  if (method !== "GET" && method !== "POST") {
    return null;
  }

  const properties = isRecord(inputSchema.properties)
    ? inputSchema.properties
    : {};
  const required = Array.isArray(inputSchema.required)
    ? inputSchema.required
    : [];

  const outputSchema = service.outputSchema || {};
  const responseProperties = isRecord(outputSchema.properties)
    ? outputSchema.properties
    : {};
  const responseFields = listResponsePaths(responseProperties);

  return {
    name: createToolName(
      String(service.serviceId || service.name || "service"),
    ),
    description: buildToolDescription(
      service,
      method,
      properties,
      required,
      responseFields,
    ),
    inputSchema: {
      type: "object",
      properties: {
        ...properties,
        responsePaths: buildResponsePathsProperty(responseFields),
      },
      required: required.filter((key) => hasOwn(properties, key)),
      additionalProperties: false,
    },
    metadata: {
      serviceId: service.serviceId,
      title: service.title || service.serviceName || service.name,
      method,
      path: service.routePath,
      authentication: {
        type: "apiKeyHeader",
        required: true,
        headerName: "X-API-Key",
        envName: "AICADE_GALAXY_API_KEY",
      },
      responseFields,
      invocationProtocol: {
        reservedFields: ["responsePaths"],
        responsePathsField: "responsePaths",
      },
    },
  };
}

function createToolName(serviceId) {
  return Array.from(serviceId, (ch) =>
    /[A-Za-z0-9_]/.test(ch) ? ch : "_",
  ).join("");
}

function buildToolDescription(
  service,
  method,
  properties,
  required,
  responseFields,
) {
  const parts = [
    service.title || service.serviceName || service.name || "AICADE service",
    service.description || "",
    method === "GET"
      ? "Use query parameters and expect a JSON response."
      : "Use a JSON request body and expect a JSON response.",
  ];

  const names = Object.keys(properties);
  if (names.length > 0) {
    parts.push(`Parameters: ${names.join(", ")}.`);

    const requiredNames = required.filter((name) => hasOwn(properties, name));
    const optionalNames = names.filter((name) => !requiredNames.includes(name));

    if (requiredNames.length > 0) {
      parts.push(`Required parameters: ${requiredNames.join(", ")}.`);
    }

    if (optionalNames.length > 0) {
      parts.push(`Optional parameters: ${optionalNames.join(", ")}.`);
    }
  }

  if (responseFields.length > 0) {
    parts.push(
      `Optional reserved field: responsePaths. Selectable response fields: ${responseFields.join(", ")}. Use responsePaths to request only the fields you need.`,
    );
  } else {
    parts.push(
      "Optional reserved field: responsePaths. Response selection is supported, but this service does not declare output fields.",
    );
  }

  return parts.filter(Boolean).join(" ");
}

function buildResponsePathsProperty(responseFields) {
  let description =
    "Optional. Select one or more response field paths to return.";
  if (responseFields.length > 0) {
    description = `${description} Available values: ${responseFields.join(", ")}.`;
  }

  return {
    type: "array",
    description,
    items: {
      type: "string",
    },
  };
}

function listResponsePaths(properties) {
  const paths = [];
  for (const [key, prop] of Object.entries(properties)) {
    walkSchemaProperty(key, prop, paths);
  }
  return paths;
}

function walkSchemaProperty(currentPath, prop, paths) {
  paths.push(currentPath);

  if (prop?.type === "object" && isRecord(prop.properties)) {
    for (const [key, child] of Object.entries(prop.properties)) {
      walkSchemaProperty(`${currentPath}.${key}`, child, paths);
    }
  }

  if (prop?.type === "array" && isRecord(prop.items)) {
    paths.push(`${currentPath}[]`);
    if (prop.items.type === "object" && isRecord(prop.items.properties)) {
      for (const [key, child] of Object.entries(prop.items.properties)) {
        walkSchemaProperty(`${currentPath}[].${key}`, child, paths);
      }
    }
  }
}

function joinUrl(baseUrl, path) {
  const normalizedPath = path.startsWith("/") ? path.slice(1) : path;
  return `${baseUrl}/${normalizedPath}`;
}

function isRecord(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function hasOwn(object, key) {
  return Object.prototype.hasOwnProperty.call(object, key);
}

function isAuthLikeError(message) {
  return (
    message.includes("401") ||
    message.includes("403") ||
    message.includes("X-API-Key") ||
    message.includes("AICADE_GALAXY_API_KEY")
  );
}

function isDebugEnabled(value) {
  if (!value) {
    return false;
  }

  return value.toLowerCase() === "true";
}

function logDebug(config, payload) {
  if (!config.debug) {
    return;
  }

  console.log(`[AICADE_GALAXY_DEBUG] ${JSON.stringify(payload, null, 2)}`);
}

function previewText(text, limit = 240) {
  const compact = text.split(/\s+/).filter(Boolean).join(" ");
  if (compact.length <= limit) {
    return compact;
  }
  return `${compact.slice(0, limit)}...`;
}

function dirname(path) {
  const normalized = path.replace(/\\/g, "/");
  const index = normalized.lastIndexOf("/");
  return index === -1 ? "." : normalized.slice(0, index) || ".";
}

function joinPath(dir, file) {
  const normalizedDir = dir.replace(/[\\/]+$/, "");
  if (!normalizedDir) {
    return file;
  }
  return `${normalizedDir}/${file}`;
}

await main();
